### PR TITLE
Enable bulk PDF of Delivery slip

### LINKED ISSUE #26250: Enable bulk PDF of Delivery slip
#### Is your feature request related to a problem?

Enable bulk for generating PDF of delivery slip. The functionnality can be generate one PDF with multiple page.

#### Describe the solution you'd like

Change the delivery slip generator pdf

#### Alternatives you've considered

N/a

#### Additional context

N/a

### PR DESCRIPTION
**Pull request by ytilotti**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Enable bulk for generating PDF of delivery slip
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #26250
| How to test?      | Download a delivery slip on the back office
| Possible impacts? | N/a

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26254)

### CODE REVIEW

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (tegbessou) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php]: Maybe before generating the PDF you can check if the `$orders_invoice_collection` is not null.
To avoid uselessly PDF generation

Comment (ytilotti) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php]: The code is based on https://github.com/PrestaShop/PrestaShop/blob/b56b7eb805cbd34fa8b3af7a0b7f65fd725b27e3/src/Adapter/PDF/CreditSlipPdfGenerator.php#L83-L99

But `$orders_invoice_collection` can't be empty. You want check this?
```
if (!$orders_invoice_collection->getResults()) {
    throw new RuntimeException($this->translator->trans('The order cannot be found within your database.', [], 'Admin.Orderscustomers.Notification'));
}
```
or
```
if (count($orders_invoice_collection) > 0) {
    throw new RuntimeException($this->translator->trans('The order cannot be found within your database.', [], 'Admin.Orderscustomers.Notification'));
}
```

I agree in my side. I don't why not check in `CreditSlipPdfGenerator`

Comment (tegbessou) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php]: Yes it that I want to check

Comment (PierreRambaud) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php]: Since the method is public, you can easily call this method without having parameters :thinking: 
```php
$generator->generatePDF([]);
```
I think you have two things to check, the first one is about $orderIds, and to be sure it's not empty, then 
make sure `$orders_invoice_collection->getResults()` have to be tested too

Comment (ytilotti) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php]: Like I said, I apply the same method from CreditSlipPdfGenerator. I will add this checks.

Comment (ytilotti) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php]: Done

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (Progi1984) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php:36]:
```suggestion
use PrestaShopCollection;
use RuntimeException;
```

**Review by Progi1984 — COMMENTED** [review]
@ytilotti Thank you for your contribution. A small suggestion for fixing PHPStan.

**Review by atomiix — APPROVED** [review]

**Comment by marionf** [comment]
Hello @ytilotti 

I don't see any change while downloading a delivery slip from order detail, order list, and delivery slips pages.
Could you please explain what behavior your PR changes and on which pages of the backoffice or frontoffice ?
You can provide screenshots if it's more easier for you ;)

**Comment by ytilotti** [comment]
Hi @marionf. This PR not create the bulk but enable the possibility to create the bulk.
I can make a proposal for a bulk, but I think it's an other PR.

**Comment by marionf** [comment]
Thank you for your answer @ytilotti 
Yes, it will be for another PR :)

**Review by matks — DISMISSED** [review]
1. PHP-cs-fixer requests a change in your PR.
2. I don't see the value of merging this into the software if it's not used 🤔  . Why should we add this change into the software ?

**Comment by ytilotti** [comment]
> 2. I don't see the value of merging this into the software if it's not used 🤔  . Why should we add this change into the software ?

There are multiple reasons:
- Prestashop is expandable.
- PrestaShop allows the implementation of bulk, so why block this feature on export?
- I created a module to enable a bulk and had to rewrite DeliverySlipPdfGenerator. What's the point?

Sorry missclick on review. There is nothing to approve? I don't understand why PHP CS FIXER is failing, I just fix my file in local, edit in github, there are no changes.

**Comment by matks** [comment]
@ytilotti We build PrestaShop to be used by 300,000 stores in the world.

So I cannot approve a PR that can be used only when people install your module 😅  . I cannot approve the merge of some code that is useful "only if you also install ylotti's module".

I maybe missing something though. Maybe @kpodemski or @PierreRambaud can help me understand what I am not seeing.

**Comment by ytilotti** [comment]
> So I cannot approve a PR that can be used only when people install your module 😅 . I cannot approve the merge of some code that is useful "only if you also install ylotti's module".

My module is not public, and it's not the problem. It was an example.
I think if we need to rewrite a class of PrestaShop, a PR is required.

If you want a PR with the bulk, we need to validate this PR.

**Comment by PrestaEdit** [comment]
I'm not against adding things as that, as it could be usefull for an another Pull Request that will add the possibility to make a bulk action with it. It's a good start to allow this.
And if we don't merge the bulk action form later, at least, merchants with dedicated module could make them. It's like adding an hook that no one use inside the core, it's usefull.

**Comment by jf-viguier** [comment]
The generatePDF has already an array in parameter, so maybe this PR can be considered as a bug fix ?
generatePDF(array $orderId)
IMO it's a current need to generate a bunch of pdf for an ecommerce point of view : for accounting, for a fiscal control, ...

**Comment by ytilotti** [comment]
Hi @marionf. Do you prefer I propose a bulk on a different PR or on this PR?

**Comment by marionf** [comment]
I think it can be done in another PR

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (saulaski) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php:64]:
```suggestion
        if (empty($orderIds)) {
```

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (saulaski) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php:65]: Detailled error messages are always helpful :)
```suggestion
            throw new MissingDataException('No order ID provided to generate delivery slip PDF');
```

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (saulaski) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php:74]: Is it an error case only if no order has been found ? or if a single one amongst the required orders has not been found ?
I suppose a missing order ID from the SQL `IN` clause wouldn't fit the initial intent, right ?

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (saulaski) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php:77]: Does it mean all that all delivery slips get generated in the same PDF file, instead of splitted accross multiple PDF files ?
Is it the initial intent here ?

**Review by saulaski — CHANGES_REQUESTED** [review]

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (ytilotti) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php:77]: All in the same file, yes.

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php
Comment (ytilotti) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php:74]: Don't understand this point. `OrderIds` can't be empty/null.

Comment (saulaski) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php:74]: The PHP array can't be empty indeed. But it doesn't mean all required invoices exist for real in database.
What is supposed to happen if there are less invoices than required ? Is it an error case or should we let it go as nominal case ?

Comment (ytilotti) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php:74]: It's so heavy to check all order on by one (by new Order...). We can compare size of array and throw an error if not exact. But what is the point? To add an information, this PR is inspired by other PDF adapter, example : https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/PDF/CreditSlipPdfGenerator.php. It's not check in CreditSlip.

Comment (saulaski) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php:74]: My point here is only about the legal value of delivery slips, and it's only a question of use case to implement.
If it's ok to generate the PDF document for any order whatever its status, then your PR is good enough. Just like the credit slip generator you've linked.
But if all the orders provided are expected to have an invoice available, then the user can tell a few are missing only after going through the entire PDF doc to check manually.

Previous version of the code was straight forward because it could only handle a single order at a time, and threw an exception if the related invoice wasn't found. It would let the user know if called in a simple loop through orders.
But your improvement might silence such behavior if the loop gets replaced with the bulk strategy.

I'm not telling you that's wrong or a mistake, but it might get confusing from a functional use case perspective.
I guess that's the main reason you've been previously asked to tell more about the use case :)

Comment (ytilotti) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php:74]: We can do a specific use case if `$orderIds == 1`. I thinks it's a better way to do like CreditSlip and the PDF class has a verification on empty data. There is no **single** answer to this problem ;).
But the problem is here, we can't do bulk on this adapter but on others it's possible.

Comment (saulaski) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php:74]: There is no single answer to the question indeed. And other adapters don't implement such strict behavior, which is unsafe IMHO.
I suppose the choice should be left to the shop configuration: strict or permissive. There are 3 solutions :
1. Select the behavior with an extra argument in service constructor in order to override the dependency injection in configuration file `generator.yml`
2. Extend this generator into a new class that will implement the strict behavior by default, and select the appropriate implementation in said configuration `generator.yml`
3. Add an extra method `generateStrictPDF` to `PDFGeneratorInterface` to apply the "strict mode", and let the caller pick up the most appriopriate method

Those options have pros and cons, and at the moment such behavior sounds a bit too sophisticated compared to the global design of the product. Perhaps later such "strict mode" will be added to all generators at once, but for now code consistency should prime.
Then please forget about my previous suggestion, and just fix code style in your PR. I will discuss the idea with core team and we'll decide then if strict mode" is required or not :)

**Comment by ytilotti** [comment]
Hi @matks, why in waiting for author? For me there is no more to do on this PR only the reflexion of @saulaski.

**Comment by PrestaEdit** [comment]
> Hi @matks, why in waiting for author? For me there is no more to do on this PR only the reflexion of @saulaski.

The "Waiting for author" was removed 6 days ago. It's not more with it.

**Comment by ytilotti** [comment]
Sorry missread. Thanks @PrestaEdit

**Comment by Progi1984** [comment]
@ytilotti Could you execute `php ./vendor/bin/php-cs-fixer fix` ? and push changes :) for fixing PHP CS Fixer Github Actions

**Comment by ghost** [comment]
Friendly remember @ytilotti

**Comment by matks** [comment]
Hi,

Since we had no news from you for more than 3 months, I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale. Feel free to reopen or open another one if you think it's still relevant.

Thank you for your contribution 👍 

Thanks!

**Pull request closed without merging**
